app/src/org/commcare/android/storage/framework/Persisted.java

```diff
@@ -98,6 +97,7 @@ public int getID() {
     }
 
     private void readVal(Field f, Object o, DataInputStream in) throws DeserializationException, IOException, IllegalAccessException {
+        synchronized (fieldOrderings) {
```

Comment:
Also note that making your suggested changes means we can't also correctly synchronize over the metadata field access since those aren't cached.